.github/workflows/generate.yml

```diff
@@ -15,8 +15,62 @@ permissions:
   contents: read
 
 jobs:
+  prepare:
+    runs-on: ubuntu-latest
+    outputs:
+      sha: ${{ steps.push.outputs.sha || steps.pr.outputs.sha }}
+      base-run: ${{ steps.main.outputs.run_id }}
+    steps:
+      - name: Harden Runner
+        uses: step-security/harden-runner@20cf305ff2072d973412fa9b1e3a4f227bda3c76 # v2.14.0
+        with:
+          egress-policy: audit
+
+      # If we are running from the main branch (a non-pull_request event), we
+      # want the latest SHA from nodejs/node
+      - id: push
+        if: ${{ github.event_name != 'pull_request' }}
+        run: |
+          SHA=$(git ls-remote https://github.com/nodejs/node.git HEAD | awk '{print $1}')
+          echo "$SHA" > commit
+          echo "sha=$SHA" >> "$GITHUB_OUTPUT"
+
+      - name: Upload metadata artifact
+        if: ${{ github.event_name != 'pull_request' }}
+        uses: actions/upload-artifact@330a01c490aca151604b8cf639adc76d48f6c5d4 # v5.0.0
+        with:
+          name: commit
+          path: commit
+
+      # If we are running from a PR (a pull_request event), we
+      # want the SHA used by the most recent `push` run
+      - name: Get latest `main` run
+        if: ${{ github.event_name == 'pull_request' }}
+        id: main
+        env:
+          BASE_SHA: ${{ github.event.pull_request.base.sha }}
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          ID=$(gh run list --repo $GITHUB_REPOSITORY -c $BASE_SHA -w 174604400 -L 1 --json databaseId --jq ".[].databaseId")
```

Comment:
Can you add an inline comment referencing where you found this? Also what .datebaseId is and where it comes from? (likely link to their API docs ref)